### PR TITLE
ONNX transformation to cast int64 constants to int32 when possible

### LINKED ISSUE #627: INT64 clamping to INT32 creates overhead while using TensorRT
### System Info

```shell
optimum: dev
CUDA: 11.3
cuDNN: 8.3.2
TensorRT: 8.4.1.5
```

### Who can help?

@JingyaHuang

### Reproduction

```python
from optimum.onnxruntime import ORTModelForSequenceClassification
from transformers import AutoTokenizer

ort_model = ORTModelForSequenceClassification.from_pretrained(
    "philschmid/tiny-bert-sst2-distilled",
    from_transformers=True,
    provider="TensorrtExecutionProvider",
)

tokenizer = AutoTokenizer.from_pretrained("philschmid/tiny-bert-sst2-distilled")
inp = tokenizer("expectations were low, actual enjoyment was high", return_tensors="pt", padding=True)

result = ort_model(**inp)
assert ort_model.providers == ["TensorrtExecutionProvider", "CUDAExecutionProvider", "CPUExecutionProvider"]
```

### Log

```
2022-12-21 09:40:48.208932701 [W:onnxruntime:Default, tensorrt_execution_provider.h:60 log] [2022-12-21 09:40:48 WARNING] external/onnx-tensorrt/onnx2trt_utils.cpp:367: Your ONNX model has been generated with INT64 weights, while TensorRT does not natively support INT64. Attempting to cast down to INT32
```

### Expected behavior

Clamping should be done before the inference.

Will work on a `convert_int64_to_int32()` function to solve it.

### PR DESCRIPTION
**Pull request by fxmarty**
As per title.

Partially fixes #627 , we need to integrate this in this CLI and document + test.

Try with:

```python
import onnx
from pathlib import Path
from optimum.onnx import model_to_int32

path = "/path/to/decoder_model.onnx"
model = onnx.load(path)

model = model_to_int32(model)

onnx.save(
    model,
    path,
    save_as_external_data=True,
    all_tensors_to_one_file=True,
    location=Path(path).name + "_data",
)

onnx.checker.check_model(path)
```

Inspect the original and transformed models "Slice" nodes.

### CODE REVIEW

**Comment by fxmarty** [comment]
Note: AFAIK this is not usable on > 2 GB models.

**Comment by fxmarty** [comment]
Should test with ONNX Runtime, it's possible there's an issue:

```
(hf-inf) fxmarty@huggingface:~/hf_internship/optimum/gptj_tiny_bool_int32$ onnxsim decoder_model_int32.onnx decoder_model_int32_sim.onnx --input-shape input_ids:1,12 attention_mask:1,12
Your model contains "Tile" ops or/and "ConstantOfShape" ops. Folding these ops can make the simplified model much larger. If it is not expected, please 
specify "--no-large-tensor" (which will lose some optimization chances)
Simplifying...
Traceback (most recent call last):
  File "/home/fxmarty/anaconda3/envs/hf-inf/bin/onnxsim", line 8, in <module>
    sys.exit(main())
  File "/home/fxmarty/anaconda3/envs/hf-inf/lib/python3.9/site-packages/onnxsim/onnx_simplifier.py", line 434, in main
    model_opt, check_ok = simplify(
  File "/home/fxmarty/anaconda3/envs/hf-inf/lib/python3.9/site-packages/onnxsim/onnx_simplifier.py", line 186, in simplify
    model_opt_bytes = C.simplify(
  File "/home/fxmarty/anaconda3/envs/hf-inf/lib/python3.9/site-packages/onnxsim/onnx_simplifier.py", line 239, in Run
    sess = rt.InferenceSession(
  File "/home/fxmarty/anaconda3/envs/hf-inf/lib/python3.9/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py", line 347, in __init__
    self._create_inference_session(providers, provider_options, disabled_optimizers)
  File "/home/fxmarty/anaconda3/envs/hf-inf/lib/python3.9/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py", line 386, in _create_inference_session
    sess = C.InferenceSession(session_options, self._model_bytes, False, self._read_config_from_model)
onnxruntime.capi.onnxruntime_pybind11_state.Fail: [ONNXRuntimeError] : 1 : FAIL : Type Error: Type parameter (Tind) of Optype (Slice) bound to different types (tensor(int32) and tensor(int64) in node (/transformer/h.0/attn/Slice_4).
```

**Comment by fxmarty** [comment]
Should be all good now, feel free to review.

File: optimum/onnxruntime/modeling_decoder.py [hunk 1]
```diff
@@ -305,6 +305,7 @@ def forward(
                 for input_name, past_key_value in zip(self.key_value_input_names, past_key_values):
                     onnx_inputs[input_name] = past_key_value.cpu().detach().numpy()
 
+            print(onnx_inputs)
```
Comment (fxmarty) [inline optimum/onnxruntime/modeling_decoder.py:308]:
```suggestion
```

File: optimum/onnx/graph_transformations.py [hunk 2]
```diff
@@ -92,7 +59,7 @@ def remove_duplicate_weights(model: ModelProto, inplace: bool = False) -> ModelP
     return model
 
 
-def replace_atenops_to_gather(model: ModelProto):
+def replace_atenops_to_gather(model: ModelProto) -> ModelProto:
```
Comment (michaelbenayoun) [inline optimum/onnx/graph_transformations.py:62]: Does this replace aten ops related to Gather to a working version?
If so I would make it more explicit in the name.

File: optimum/onnx/graph_transformations.py [hunk 3]
```diff
@@ -296,3 +168,41 @@ def merge_decoders(
     onnx.checker.check_model(merged_model)
 
     return merged_model
+
+
+def model_to_int32(model: ModelProto) -> ModelProto:
```
Comment (michaelbenayoun) [inline optimum/onnx/graph_transformations.py:173]: Maybe name the function to something more explicit:
```suggestion
def slice_arguments_to_int32(model: ModelProto) -> ModelProto:
```
This is only a suggestion but I think that the current name is not clear enough.

**Review by michaelbenayoun — APPROVED** [review]
LGTM

File: optimum/onnx/graph_transformations.py (hunk 2 above)
Comment (fxmarty) [reply optimum/onnx/graph_transformations.py:62]: No idea, I did not code it.

**Pull request merged**